i-beam cursor switches to black square when typing

I have been running latest xrdp and xorgxrdp 0.9 for quite some time.
I compile from source.  I'm running Rocky Enterprise Linux 8.10 (RHEL8.10).
Today,  I updated to xorgxrdp 0.10.2 and xrdp 0.10.1 and am having one issue.

When I login to my XRDP session, my gnome session starts.  I open a terminal.   If I hit backspace, I hear the beep (because I've configured sound) and immediately the mouse cursor switches from the (I think it's called) "I-beam" to just a black square!   I can get it to switch back to the "I-beam" again by just moving the mouse.
If I start typing in the Terminal window, the cursor switches back to the black square, and again, if I move the mouse it switches back to the I-Beam.
Now, if I click on the terminal so that my text is going to go in there, and move the mouse cursor out of the terminal Window so that it turns to the arrow, then I can type, and my text will go into the window, and the cursor will not change...
So this seems to impact the I-beam cursor only...  But why? This has never been an issue for me in the past.

I have this installed on > 150 systems, and the same experience on all of them (unfortunately).

(I believe this is an xorgxrdp issue... My apologies if this is an xrdp issue).

I should add that I viewed the xrdp log and see some errors that could be relevant:

```
[2024-08-27T15:58:51.685-0400] [INFO ] starting xrdp with pid 10427
[2024-08-27T15:58:51.686-0400] [INFO ] address [0.0.0.0] port [3389] mode 1
[2024-08-27T15:58:51.687-0400] [INFO ] listening to port 3389 on 0.0.0.0
[2024-08-27T15:58:51.688-0400] [INFO ] xrdp_listen_pp done
[2024-08-27T15:58:58.359-0400] [INFO ] Socket 14: connection accepted from 130.63.94.85:36454
[2024-08-27T15:58:58.361-0400] [INFO ] Using default X.509 certificate: /xconf/xrdp/cert.pem
[2024-08-27T15:58:58.362-0400] [INFO ] Using default X.509 key file: /xconf/xrdp/key.pem
[2024-08-27T15:58:58.363-0400] [INFO ] Security protocol: configured [SSL|RDP], requested [SSL|HYBRID|RDP], selected [SSL]
[2024-08-27T15:58:58.368-0400] [INFO ] Connected client computer name: Guacamole RDP
[2024-08-27T15:58:58.369-0400] [WARN ] Physical desktop dimensions (0x0) are invalid
[2024-08-27T15:58:58.370-0400] [WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc006 is unknown (ignored)
[2024-08-27T15:58:58.371-0400] [WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc00a is unknown (ignored)
[2024-08-27T15:58:58.968-0400] [INFO ] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x04], keyboard_subtype:[0x00], keylayout:[0x00000409]
[2024-08-27T15:58:58.970-0400] [INFO ] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[2024-08-27T15:58:58.971-0400] [INFO ] TLS connection established from my-ip:36454 TLSv1.3 with cipher TLS_AES_256_GCM_SHA384
[2024-08-27T15:58:59.169-0400] [INFO ] Client supports single-screen resizes by xrdp
[2024-08-27T15:58:59.170-0400] [INFO ] xrdp_caps_process_pointer: client supports new(color) cursor
[2024-08-27T15:58:59.171-0400] [INFO ] xrdp_process_offscreen_bmpcache: support level 1 cache size 7864320 MB cache entries 2000
[2024-08-27T15:58:59.172-0400] [WARN ] Client Capability: not enough orders supported by client, client wants off screen bitmap but offscreen bitmaps disabled
[2024-08-27T15:58:59.173-0400] [INFO ] Client Capability: LARGE_POINTER_FLAG_96x96 supported
[2024-08-27T15:58:59.177-0400] [INFO ] Loading keymap file /xconf/xrdp/km-00000409.ini
[2024-08-27T15:58:59.179-0400] [WARN ] local keymap file for 0x00000409 found and doesn't match built in keymap, using local keymap file
[2024-08-27T15:58:59.180-0400] [WARN ] Static channel 'drdynvc' not found.
[2024-08-27T15:58:59.181-0400] [WARN ] Dynamic channels will not be available
[2024-08-27T15:58:59.182-0400] [WARN ] No information is available to determine login screen DPI
[2024-08-27T15:58:59.183-0400] [WARN ] No DPI value is available to find login font
[2024-08-27T15:58:59.184-0400] [WARN ] Using the default_dpi of 96
[2024-08-27T15:58:59.187-0400] [WARN ] Ignoring obsolete SCP port value '3350'
[2024-08-27T15:58:59.189-0400] [INFO ] xrdp_wm_log_msg: sesman connect ok
[2024-08-27T15:58:59.190-0400] [INFO ] xrdp_wm_log_msg: Performing login request for jas
[2024-08-27T15:58:59.422-0400] [INFO ] xrdp_wm_log_msg: login was successful - creating session
[2024-08-27T15:58:59.654-0400] [ERROR] xrdp_rdp_recv: xrdp_channel_process failed
[2024-08-27T15:59:00.515-0400] [INFO ] xrdp_wm_log_msg: session is available on display 11 for user jas
[2024-08-27T15:59:00.516-0400] [INFO ] xrdp_wm_log_msg: Connecting to session
[2024-08-27T15:59:00.517-0400] [INFO ] loaded module 'libxup.so' ok, interface size 10416, version 4
[2024-08-27T15:59:00.518-0400] [INFO ] lib_mod_connect: connecting via UNIX socket
[2024-08-27T15:59:00.519-0400] [INFO ] lib_mod_log_peer: xrdp_pid=10475 connected to Xorg_pid=10479 Xorg_uid=1004 Xorg_gid=1000 client=130.63.94.85:36454
[2024-08-27T15:59:00.519-0400] [INFO ] xrdp_wm_log_msg: Connecting to chansrv
[2024-08-27T15:59:02.525-0400] [INFO ] Received memory_allocation_complete command. width: 2560, height: 1271
```
In particular, the "Client Capability: not enough orders supported by client, client wants off screen bitmap but offscreen bitmaps disabled".

In the previous xrdp log none of those messages were there...

Jason.

--- a/module/rdpClientCon.c
+++ b/module/rdpClientCon.c
@@ -2109,6 +2109,27 @@ rdpClientConDrawLine(rdpPtr dev, rdpClientCon *clientCon,
 
 /******************************************************************************/
 int
+rdpClientConSetCursorSystem(rdpPtr dev, rdpClientCon *clientCon,
+                            int pointer_type)
+{
+    int size;
+
+    if (clientCon->connected)
+    {
+        LLOGLN(10, ("rdpClientConSetCursor:"));
+        size = 2 + 2 + 4;
+        rdpClientConPreCheck(dev, clientCon, size);
+        out_uint16_le(clientCon->out_s, 65); /* set cursor system */
+        out_uint16_le(clientCon->out_s, size); /* size */
+        clientCon->count++;
+        out_uint32_le(clientCon->out_s, pointer_type);
+    }
+
+    return 0;
+}
+
+/******************************************************************************/
+int
 rdpClientConSetCursor(rdpPtr dev, rdpClientCon *clientCon,
                       short x, short y, uint8_t *cur_data, uint8_t *cur_mask)
 {

--- a/module/rdpClientCon.c
+++ b/module/rdpClientCon.c
@@ -2130,6 +2130,27 @@ rdpClientConSetCursorSystem(rdpPtr dev, rdpClientCon *clientCon,
 
 /******************************************************************************/
 int
+rdpClientConMoveCursor(rdpPtr dev, rdpClientCon *clientCon, int x, int y)
+{
+    int size;
+
+    if (clientCon->connected)
+    {
+        LLOGLN(10, ("rdpClientConSetCursor:"));
+        size = 2 + 2 + 2 + 2;
+        rdpClientConPreCheck(dev, clientCon, size);
+        out_uint16_le(clientCon->out_s, 66); /* move cursor */
+        out_uint16_le(clientCon->out_s, size); /* size */
+        clientCon->count++;
+        out_uint16_le(clientCon->out_s, x);
+        out_uint16_le(clientCon->out_s, y);
+    }
+
+    return 0;
+}
+
+/******************************************************************************/
+int
 rdpClientConSetCursor(rdpPtr dev, rdpClientCon *clientCon,
                       short x, short y, uint8_t *cur_data, uint8_t *cur_mask)
 {

--- a/module/rdpClientCon.h
+++ b/module/rdpClientCon.h
@@ -181,6 +181,8 @@ extern _X_EXPORT int
 rdpClientConSetCursorSystem(rdpPtr dev, rdpClientCon *clientCon,
                             int pointer_type);
 extern _X_EXPORT int
+rdpClientConMoveCursor(rdpPtr dev, rdpClientCon *clientCon, int x, int y);
+extern _X_EXPORT int
 rdpClientConSetCursor(rdpPtr dev, rdpClientCon *clientCon,
                       short x, short y, uint8_t *cur_data, uint8_t *cur_mask);
 extern _X_EXPORT int

--- a/module/rdpClientCon.h
+++ b/module/rdpClientCon.h
@@ -178,6 +178,9 @@ rdpClientConAddAllReg(rdpPtr dev, RegionPtr reg, DrawablePtr pDrawable);
 extern _X_EXPORT int
 rdpClientConAddAllBox(rdpPtr dev, BoxPtr box, DrawablePtr pDrawable);
 extern _X_EXPORT int
+rdpClientConSetCursorSystem(rdpPtr dev, rdpClientCon *clientCon,
+                            int pointer_type);
+extern _X_EXPORT int
 rdpClientConSetCursor(rdpPtr dev, rdpClientCon *clientCon,
                       short x, short y, uint8_t *cur_data, uint8_t *cur_mask);
 extern _X_EXPORT int

--- a/module/rdpCursor.c
+++ b/module/rdpCursor.c
@@ -251,6 +251,7 @@ rdpSpriteSetCursorCon(rdpClientCon *clientCon,
     int sending_bpp;
     int can_do_new;
     int can_do_large;
+    int cursor_id;
 
     LLOGLN(10, ("rdpSpriteSetCursorCon:"));
     if (clientCon->suppress_output)
@@ -281,12 +282,13 @@ rdpSpriteSetCursorCon(rdpClientCon *clientCon,
     if ((pCurs == NULL) || (pCurs->bits == NULL))
     {
         /* None cursor */
-        sending_width = 32;
-        sending_height = 32;
-        xhot = 0;
-        yhot = 0;
-        memset(cur_data, 0, 96 * 96 * 4);
-        memset(cur_mask, 0xFF, 96 * 96 / 8);
+        cursor_id = 0; /* SYSPTR_NULL */
+        LLOGLN(0, ("rdpSpriteSetCursorCon: sending cursor system "
+                "pointer 0x%8.8X", cursor_id));
+        rdpClientConBeginUpdate(clientCon->dev, clientCon);
+        rdpClientConSetCursorSystem(clientCon->dev, clientCon, cursor_id);
+        rdpClientConEndUpdate(clientCon->dev, clientCon);
+        return;
     }
     else
     {

--- a/module/rdpCursor.c
+++ b/module/rdpCursor.c
@@ -225,10 +225,10 @@ set_pixel_safe(uint8_t *data, int x, int y, int width, int height, int bpp,
 }
 
 /******************************************************************************/
-void
+static void
 rdpSpriteSetCursorCon(rdpClientCon *clientCon,
                       DeviceIntPtr pDev, ScreenPtr pScr, CursorPtr pCurs,
-                      int x, int y)
+                      int x, int y, int do_move)
 {
     uint8_t *cur_data;
     uint8_t *cur_mask;
@@ -283,7 +283,7 @@ rdpSpriteSetCursorCon(rdpClientCon *clientCon,
     {
         /* None cursor */
         cursor_id = 0; /* SYSPTR_NULL */
-        LLOGLN(0, ("rdpSpriteSetCursorCon: sending cursor system "
+        LLOGLN(10, ("rdpSpriteSetCursorCon: sending cursor system "
                 "pointer 0x%8.8X", cursor_id));
         rdpClientConBeginUpdate(clientCon->dev, clientCon);
         rdpClientConSetCursorSystem(clientCon->dev, clientCon, cursor_id);
@@ -388,6 +388,10 @@ rdpSpriteSetCursorCon(rdpClientCon *clientCon,
                                    cur_data, cur_mask, sending_bpp,
                                    sending_width, sending_height);
     }
+    if (do_move)
+    {
+        rdpClientConMoveCursor(clientCon->dev, clientCon, x, y);
+    }
     rdpClientConEndUpdate(clientCon->dev, clientCon);
     free(cur_data);
 }
@@ -399,22 +403,57 @@ rdpSpriteSetCursor(DeviceIntPtr pDev, ScreenPtr pScr, CursorPtr pCurs,
 {
     rdpPtr dev;
     rdpClientCon *clientCon;
+    int do_move;
 
     LLOGLN(10, ("rdpSpriteSetCursor:"));
     dev = rdpGetDevFromScreen(pScr);
+    LLOGLN(10, ("rdpSpriteSetCursor: x %d y %d cursor_x %d cursor_y %d",
+           x, y, dev->pointer.cursor_x, dev->pointer.cursor_y));
+    do_move = (dev->pointer.cursor_x != x) || (dev->pointer.cursor_y != y);
+    LLOGLN(10, ("rdpSpriteSetCursor: x %d y %d do_move %d", x, y, do_move));
     clientCon = dev->clientConHead;
     while (clientCon != NULL)
     {
-        rdpSpriteSetCursorCon(clientCon, pDev, pScr, pCurs, x, y);
+        rdpSpriteSetCursorCon(clientCon, pDev, pScr, pCurs, x, y, do_move);
         clientCon = clientCon->next;
     }
+}
+
+/******************************************************************************/
+static void
+rdpSpriteMoveCursorCon(rdpClientCon *clientCon,
+                      DeviceIntPtr pDev, ScreenPtr pScr, int x, int y)
+{
+    LLOGLN(10, ("rdpSpriteMoveCursorCon:"));
+    rdpClientConBeginUpdate(clientCon->dev, clientCon);
+    rdpClientConMoveCursor(clientCon->dev, clientCon, x, y);
+    rdpClientConEndUpdate(clientCon->dev, clientCon);
 }
 
 /******************************************************************************/
 void
 rdpSpriteMoveCursor(DeviceIntPtr pDev, ScreenPtr pScr, int x, int y)
 {
+    rdpPtr dev;
+    rdpClientCon *clientCon;
+    int do_move;
+
     LLOGLN(10, ("rdpSpriteMoveCursor:"));
+    dev = rdpGetDevFromScreen(pScr);
+    LLOGLN(10, ("rdpSpriteMoveCursor: x %d y %d cursor_x %d cursor_y %d",
+      x, y, dev->pointer.cursor_x, dev->pointer.cursor_y));
+    do_move = (dev->pointer.cursor_x != x) || (dev->pointer.cursor_y != y);
+    LLOGLN(10, ("rdpSpriteMoveCursor: x %d y %d do_move %d", x, y, do_move));
+    if (!do_move)
+    {
+        return;
+    }
+    clientCon = dev->clientConHead;
+    while (clientCon != NULL)
+    {
+        rdpSpriteMoveCursorCon(clientCon, pDev, pScr, x, y);
+        clientCon = clientCon->next;
+    }
 }
 
 /******************************************************************************/

